storage/filesystem/internal/dotgit/dotgit.go

```diff
@@ -251,13 +284,30 @@ func (d *DotGit) SetRef(r *plumbing.Reference) error {
 		content = fmt.Sprintln(r.Hash().String())
 	}
 
-	f, err := d.fs.Create(r.Name().String())
+	// If we are not checking an old ref, just truncate the file.
+	mode := os.O_RDWR | os.O_CREATE
+	if old == nil {
+		mode |= os.O_TRUNC
+	}
+
+	f, err := d.fs.OpenFile(r.Name().String(), mode, 0666)
 	if err != nil {
 		return err
 	}
 
 	defer ioutil.CheckClose(f, &err)
 
+	err = f.Lock()
```

Comment:
Shouldn't there be a corresponding `f.Unlock()` somewhere, maybe as a `defer`?

Reply:
Actually Close takes care of it, but should probably document that. (If we were unlocking the file it would probably(?) need a fsync before the unlock, but go-billy lacks fsync too).

Reply:
@taruti: I don't understand how `Close()` takes care of it.  Is that a new requirement of the `billy.File.Close()` method?  Our current implementation doesn't do that, though I guess maybe for posfix flock and Windows, Close(), handles it automatically?  We will need something that explicitly calls `Unlock()` on our own `libfs.File` implementation; if `libfs.File.Close()` is supposed to do that, we can make that happen.  Was just confused.  A comment here would definitely help, and maybe one over in `go-billy` as well if that's a new assumption.

Reply:
But flocks are released by closing the file? From e.g. Linux ```Furthermore, the lock is released either by an explicit LOCK_UN operation on any of these duplicate file descriptors, or when all such file descriptors have been closed.```. But yes more documentation would be good.

Reply:
Our implementation of file locking will run into trouble with other applications if we keep the locks after all file descriptors are closed.

Reply:
The implementation of our KBFS locks won't be flock-based, so we'll need some way to know to release them.  Doing it as part of the `Close()` call is totally fine, we should just document it.

(@songgao -- we should change our `libfs.File.Close()` method to call `Unlock()`).